samples/xtensa_asm2/src/main.c

```diff
@@ -0,0 +1,539 @@
+		 * interrupted and restored, and validate that it
+		 * returns the same result every time.  But this can't
+		 * work, even in principle: the timer interrupt we are
+		 * using is "high priority", which means that it can
+		 * interrupt the window exceptions being thrown in the
+		 * main thread.  And by design, Xtensa window
+		 * exceptions CANNOT be made reentrant (they don't
+		 * save the interrupted state, so can be interrupted
+		 * again before they can mask off exceptions, which
+		 * will then lose/clobber the OWB field in PS when the
+		 * interrupt handler throws another window exception).
+		 * So this doesn't work, in fact it fails every 2-10
+		 * iterations as spill_fn spends a lot of its time
+		 * spill/filling stack frames (by design, of course).
+		 *
+		 * This could be made to work if we could repurpose
+		 * the existing medium priority timer interrupt (which
+		 * is hard in a unit test: that's an important
+		 * interrupt!) or use the low priority timer which
+		 * delivers to the global exception handler (basically
+		 * impossible in a unit test).  Frustrating.
+		 */
+		int reps = 0;
+
+		while (!timer2_fired && reps < (max_reps+2)) {
+			int result = spill_fnp(0, 3, 2, 1);
+
+			reps++;
+			if (result != expect) {
+				ok = 0;
+			}
+		}
+		if (reps >= max_reps+2) {
+			printk("Interrupt didn't arrive\n");
+			ok = 0;
+		}
+		if (int5_result != expect) {
+			printk("Unexpected int spill_fn() result\n");
+			ok = 0;
+		}
+		printk("INT test delay %d nest %d reps %d\n",
+		       delay, nest, reps);
+#else
+		/* So this is what we do instead: just spin in the
+		 * main thread calling functions that don't involve
+		 * exceptions.  By experiment, calling spill_fn with a
+		 * first (depth) argument of 6 or 7 results in a
+		 * shallow call tree that won't throw exepctions.  At
+		 * least we're executing real code which depends on
+		 * its register state and validating that interrupts
+		 * don't hurt.
+		 */
+		volatile int dummy = 1;
+
+		while (!timer2_fired) {
+			dummy = spill_fnp(6, dummy, 2, 3);
+		}
+		if (int5_result != expect) {
+			printk("Unexpected int spill_fn() result\n");
+			ok = 0;
+		}
+#endif
+	}
+	return ok;
+}
+
+void main(void)
+{
+	/* Turn off interrupts and leave disabled, otherwise the
+	 * "userspace" context switching tests might not be reliable.
+	 * Stack pointers can exist in indeterminate states here.
+	 * (Note: the interrupt test below is using a high priority
+	 * interrupt which is not masked by irq_lock(), so it doesn't
+	 * care).
+	 */
+	int key = irq_lock();
+
+	/* Strictly not a "test", we just want to know that the jump
+	 * worked.  If the rest of the code runs, this must have
+	 * "passed".
+	 */
+	rfi_jump();
+
+	int ok = 1;
+
+	ok = ok && test_reg_spill();
+
+	ok = ok && test_highreg_save();
+
+	ok = ok && test_switch();
+
+	ok = ok && test_xstack();
+
+	ok = ok && interrupt_test();
+
+	irq_unlock(key);
+
+	printk("%s\n", ok ? "OK" : "Failed");
```

Comment:
please convert this to ztest so that success/failure is reported in the expected way and move this under tests/.
We want sanitycheck/test case framework to exercise this in qemu-xtensa or real hardware, since we appear to be able to determine success/failure at runtime.

Reply:
agree